ld-floxlib/unit-test.c

```diff
@@ -0,0 +1,59 @@
+/*
+ * Unit test runner for ld-floxlib.so.
+ *
+ * By design ld-floxlib only parses FLOX_ENV_LIB_DIRS once per invocation,
+ * so testing la_objsearch() for various combinations of env and arg input
+ * requires multiple invocations. This test program calls the ld-floxlib.so
+ * la_objsearch() function with the provided "name" arg and asserts that it
+ * returns the expected value. It also performs a quick test of la_version()
+ * for good measure while we're in the neighbourhood.
+ */
+
+#ifndef _GNU_SOURCE
+#  define _GNU_SOURCE
+#endif /* _GNU_SOURCE */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <link.h>
+extern unsigned int la_version( unsigned int version );
+extern char * la_objsearch( const char * name, uintptr_t * cookie, unsigned int flag );
+
+int
+main( int argc, char **argv )
+{
+  if ( argc != 3 ) {
+    fprintf(stderr, "ERROR: expected 2 arguments, received %d\n"
+      "USAGE: %s <name_to_lookup> <expected_value>\n", (argc - 1), argv[0]);
+    exit(1);
+  }
+
+  // la_version() is basically the identity function. It's worth the
+  // usecond or two to give la_version() a quick test.
+  assert(la_version(1) == 1);
+  assert(la_version(2) == 2);
+  assert(la_version(3) != 2);
+  assert(la_version(-1) == -1);
+
+  // la_objsearch() searches the contents of the FLOX_ENV_LIB_DIRS
+  // variable looking for library matches, but only when invoked
+  // with the LA_SER_DEFAULT flag. Take a moment to ensure all other
+  // flags return the input unaltered.
+  assert(la_objsearch(argv[1], (uintptr_t *) NULL, LA_SER_ORIG   ) == argv[1]);
+  assert(la_objsearch(argv[1], (uintptr_t *) NULL, LA_SER_LIBPATH) == argv[1]);
+  assert(la_objsearch(argv[1], (uintptr_t *) NULL, LA_SER_RUNPATH) == argv[1]);
+  assert(la_objsearch(argv[1], (uintptr_t *) NULL, LA_SER_CONFIG ) == argv[1]);
+  assert(la_objsearch(argv[1], (uintptr_t *) NULL, LA_SER_SECURE ) == argv[1]);
```

Comment:
nonblocking suggestion: I think these checks are performed 7x which isn't really necessary. Not a huge deal since they're so trivial, but I think it would make more sense if they only ran once.

Reply:
Yeah the `la_version()` checks are performed 7x but they also complete in nanoseconds. The non-LA_SER_DEFAULT checks are also performed 7x, but with a different `name` value each time, which I thought might add value?

In truth we can probably just get rid of lines 35-48 altogether as they all represent the simplest functionality possible, which is just to return the same name that was passed to each of the `la_version()` and `la_objsearch()` functions. Would you prefer that we do that? I wouldn't mind honestly.